### ISSUE TITLE
[DTensor] Flaky ops test related to incorrect sharding prop caching for nested native ops

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#132114]
**Overview**
The following three DTensor op tests are marked flaky in CI:
- `dstack`: https://github.com/pytorch/pytorch/issues/126493
- `inner`: https://github.com/pytorch/pytorch/issues/126852
- `vstack`: https://github.com/pytorch/pytorch/issues/126868

Running them locally as part of the entire op test reproduces the failure:
```
pytest test/distributed/_tensor/test_dtensor_ops.py -k test_dtensor_op_db 
```
Running them individually passes.

**Investigation**
After some looking, this appears to be due to an issue with the sharding propagation caching, where these ops are using incorrect cached sharding results.
- Evidence 1: Disable the `lru_cache` for the sharding propagation, and the failures go away:
https://github.com/pytorch/pytorch/blob/aa1488fe02a350874d470348148649fc81bac302/torch/distributed/_tensor/_sharding_prop.py#L49
- Evidence 2: Specially do not use the cached sharding propagation for `unsqueeze` and `view`:
```
elif "unsqueeze" in str(op_info.schema.op) or "view" in str(op_info.schema.op):
    output_sharding = self.propagate_op_sharding_non_cached(op_info.schema)
```
Avoiding `unsqueeze` caching fixes the failures for `dstack` and `vstack`. Avoiding `view` caching fixes the failure for `inner`.

I need some help for more details, but what is special about these ops? Their native implementations call into other native ops.
- `dstack` -> `atleast_3d` -> `unsqueeze`
- `vstack` -> `atleast_2d` -> `unsqueeze`
- `inner` -> `view`

So, it appears that the hashing is somehow not taking into these into account, leading to incorrect results.

cc @wanchaol @XilunWu @chauhang @d4l3k @tianyu-l @wz337 @bdhirsh

**Comment by XilunWu** [comment]
did a print debugging and found weird stuff:
```
vstack([
    [a1, a2, a3, a4],
    [b1, b2, b3, b4],
    [c1, c2, c3, c4],
])
```
with placements of each tensor above being `[Replicate()]`, `[Replicate()]`, `[Shard(0)]` correspondingly.

the `unsqueeze` step produces correct result:
`[a1, a2, a3, a4]` -> `[[a1, a2, a3, a4]]` with placements being `[Replicate()]`
`[b1, b2, b3, b4]` -> `[[b1, b2, b3, b4]]` with placements being `[Replicate()]`
`[c1, c2, c3, c4]` -> `[[c1, c2, c3 ,c4]]` sharded on tensor dim-1. i.e. rank 0 has `[[c1]]`.

However, at the `cat` step something goes wrong:
tensor A and B are wrongly sharded in a way that rank 0 has `[[a4]]`, `[[b4]]`; and rank 1 has `[[a1]]`, `[[b1]]`, etc.

So far I'm confused why turning off sharding prop cache would fix this.

**Comment by XilunWu** [comment]
Root cause: sharding propagator cache is shared among threads but isolated among processes. In Multi-threaded PG testing, this shared cache gives wrong sharding result.

fix: #134294 makes sharding propagator not shared among all threads and should fix these 3 tests.
